hash.go

```diff
@@ -38,6 +39,57 @@ func TreeHash(val interface{}) ([32]byte, error) {
 	return paddedOutput, nil
 }
 
+// SigningRoot returns the signed message for the value passed after truncating
+// the `signature` field.
+func SigningRoot(val interface{}) ([32]byte, error) {
+	valObj := reflect.ValueOf(val)
+	kind := valObj.Kind()
+
+	switch {
+	case kind == reflect.Struct:
+		truncated, err := truncateLast(valObj.Type())
```

Comment:
paths for both types (struct and pointer) are close to identical, they should be abstracted away into a common function